Improve derridaean_similarity using quantum simulation

Qubitium opened the pull request:
derridaean_similarity is pretty damn good, after testing, but its output is entirely predicated on the quality of the random number generation.

This PR proposes an optional quantum toggle that will enable quantum random number generation using ~~8-qubit~~ 1-qubit simulated system to provide quantifiably more robust random number output.

I did some benchmarks and quantumfied derridaean_similarity is approximately 3.5m (m as in million) times slower than non-quantum but I believe the results speak for themselves. Due to the speed cost, I have set it to default to false.

p.s. Please accept my apology for premature rant in issue https://github.com/jdagdelen/hyperDB/issues/2. I have come to my senses. I wish to contribute to the hypeDB moving forward.

@rjadr @jdagdelen @jobergum

diegomontoya committed:
Add optional quantum random generator to improve the accuracy of idaean_similarity where random number generation is positively the only factor affecting output

Misc: cleanup python formatting according to PEP rules in pycharm and fix path issues demo.py

Qubitium commented:
Pretty sure the performance penalty can be reduced by several million factors using 2-qubit system vs current default of 8-qubits. But I am not skilled enough to change the qiskit backend to simulate 2-qubits instead of 8.

jdagdelen commented:
I love this, but it adds another requirement to the package and we are tying to avoid that at this time and remain pure numpy. 

While HyperDB is a joke, it is also not a joke. We do actually want it to be a useful library that helps people avoid writing the same boilerplate when developing knowledge augmented LLM applications.

diegomontoya committed:
1) Remove qiskit and qrng dependency which are not MIT compatible
2) Use a single qubit (simulated) to generate random from -0.2 to 0.2
3) 3 millions times faster than qiskit version
4) quantum random as default for derridaean_similarity

Qubitium commented:
@jdagdelen  Thanks for the feedback.  All your concerns have been resolved in the latest pr update. 

1) Remove qiskit and qrng dependency which are not MIT compatible
2) Use a single qubit (simulated) to generate quantum-based random
3) 3 millions times faster than qiskit version
4) quantum random as default for derridaean_similarity

diegomontoya committed:
replace string concat with bitwise ops

jdagdelen reviewed (commented):
Can you please move Qubit class definition outside of the similarity metric if possible? I think we should actually move all derridean similarity stuff to a separate file and add some doctorings at the top to explain what it is.

diegomontoya committed:
1. Move qubit class outside
2. Make qubit class thread safe and scalable to HPC cluster (vcs dig this)
3. Note that this code is export restricted (due to quantum) (vcs will eat this up)
4. Now even faster than before! (vss doesn't care about this unfortunately)
5. comments for clarity

Qubitium commented:
1. Move qubit class outside (now shareable for others to use!)
2. Make qubit class thread safe and scalable to HPC cluster (vcs dig this)
3. Note that this code is export restricted (due to quantum) (vcs will eat this up)
4. Now even faster than before! (vss doesn't care about this unfortunately)
5. comments for clarity

rjadr commented:
I'm not sure if we realize how much we're innovating poststructural semiotics here. This commit moves way beyond the deconstruction of LLMs' logocentrism. Linking the ultimate indeterminacy of meaning to quantum indeterminacy will allow for a reconceptualisation of randomness as a necessary condition for communication rather than a limitation or a source of noise.
Perhaps we should write a paper about it? I already have a title: "Transgressing the Boundaries: Towards a Transformative Hermeneutics of Quantum Semiotics"

jdagdelen commented:
Ok, let's add some unit tests since this is getting a bit more complex, but this looks good.

Qubitium commented:
@rjadr Can you do the derridean sim unit tests  on this PR? Need your expertise on derridean. I am very much unqualified to check if the quantum simulation enhances or breaks the original principle.
